### DISCUSSION TITLE
Using TypeSchema adapters?

### DISCUSSION THREAD

**Discussion by decs** [jellydn/next-validations#824] [Ideas]
hi, @jellydn!

I maintain a project called [TypeSchema](https://github.com/decs/typeschema), which is an universal adapter for all validation libraries (17 and counting!). `next-validations` could benefit from delegating the adapter part to TypeSchema. it would enable support for many more libraries with no extra effort, and make things easier to maintain on your side. would something like that seem like a good idea?

**Comment by jellydn** [comment]
Hi, @decs Thank you for bringing up this topic. I'm not quite sure about your suggestion. Are you saying that next-validations could use your package as an internal tool and would support more validation libraries? If so, that's a good idea. Could you open a PR about it?

**Reply by decs** [reply]
that's right. it would be an internal dependency to next-validations. then any end developers using next-validations would be able to pass any validation schema and it would work as expected

I can work on a PR to show that integration. there are 2 ways: (from the Quickstart guide)
> **Handpicking adapters** with [`@typeschema/main`](https://github.com/decs/typeschema/tree/main/packages/main): Recommended for library maintainers. Any validation library can be used, but adapters have to be explicitly installed. This allows end developers to trade-off between coverage and bundle size.
> **Batteries included** with [`@typeschema/all`](https://github.com/decs/typeschema/tree/main/packages/all): Easiest to use. All adapters are automatically installed, including future ones. This is a drop-in replacement for the deprecated [`@decs/typeschema`](https://www.npmjs.com/package/@decs/typeschema).

my recommendation is using `@typeschema/main`. if an end developer wants to use arktype schemas, for example, they just have to add `@typeschema/arktype` (together with `next-validation`) as a dependency on their project. otherwise, they'd see a runtime error saying that they need to install `@typeschema/arktype`. does that experience sound good?

another alternative is using `@typeschema/all`, which already includes all adapters (so, a simpler developer experience), but would result in a larger bundle size to end developers

**Reply by jellydn** [reply]
I would prefer the 1st option. The developer will install the package if they need anything.

**Reply by decs** [reply]
sent PR on #825. we can continue the discussion there. one caveat I noticed is that TypeSchema doesn't support fastest-validator yet, but I can prioritize that if it's a blocker

**Reply by jellydn** [reply]
Great. Could you fix the conflict on the PR? https://github.com/jellydn/next-validations/pull/825

> support fastest-validator

It's not a blocker but it would be good if TypeSchema could support it soon :) Thank you for your quick action on this idea.

**Reply by decs** [reply]
solved the conflict. could you help handle the README changes and the version publishing? feel free to modify the PR as you wish

**Reply by jellydn** [reply]
Hi @decs I have just tested with the pre-release for the next-validations. Looks good.

Could we support https://www.npmjs.com/package/fastest-validator as well? That's the final part then I could publish ver 1.0.0. Thank you very much.

**Reply by decs** [reply]
done! merged https://github.com/decs/typeschema/pull/46 to add support for `fastest-validator`
you can update `@typeschema/main` to `v0.13.8` and we're good to go

**Reply by jellydn** [reply]
Hi @decs I've got the TS error on this https://github.com/jellydn/next-validations/commit/09f27623d2eacfaa495bef038bfbc1e391e4d4c7 Could you take a look? Thanks.

```
Diagnostics:
Type '{ name: { type: string; min: number; max: number; }; email: { type: string; }; age: string; }' is not assignable to type 'AnySchema<any> | BaseSchema<any, any> | BaseSchemaAsync<any, any> | Schema<any, any, any, ""> | ZodType<any, ZodTypeDef, any>'.

TypeScript Error Translation(s):
  • I was expecting a type matching 'AnySchema<any> | BaseSchema<any, any> | BaseSchemaAsync<any, any> | Schema<any, any, any, ""> | ZodType<any, ZodTypeDef, any>', but instead you passed '{ name: { type: string; min: number; max: number; }; email: { type: string; }; age: string; }'.
```


```ts
import type { NextApiRequest, NextApiResponse } from 'next';
import { withValidation } from 'next-validations';

// This is fastest-validator schema
const schema = {
  name: { type: 'string', min: 3, max: 255 },
  email: { type: 'email' },
  age: 'number',
};

const validate = withValidation({
  schema,
  mode: 'body',
});

const handler = (req: NextApiRequest, res: NextApiResponse) => {
  res.status(200).json(req.body);
};

export default validate(handler);
```

**Reply by decs** [reply]
We also need to upgrade `@typeschema/main` to `v0.13.8` so that the fastest-validator adapter gets checked

**Reply by jellydn** [reply]
I've already updated @decs refer https://github.com/jellydn/next-validations/blob/main/package.json#L65

**Reply by decs** [reply]
I tried with a clean install of next-validations and that piece of code worked fine. are you still seeing this error? could you try with a clean git clone?
